src/gridfs.rs

```diff
@@ -0,0 +1,274 @@
+            if let Some(ref wc) = options.write_concern {
+                return Some(wc);
+            }
+        }
+        self.db.write_concern()
+    }
+
+    /// Gets the read preference of the [`GridFsBucket`].
+    pub fn read_preference(&self) -> Option<&ReadPreference> {
+        if let Some(options) = &self.options {
+            if let Some(ref rp) = options.read_preference {
+                return Some(rp);
+            }
+        }
+        if let Some(SelectionCriteria::ReadPreference(ref rp)) = self.db.selection_criteria() {
+            Some(rp)
+        } else {
+            None
+        }
+    }
+
+    /// Opens a [`GridFsStream`] that the application can write the contents of the file to.
+    /// The application provides a custom file id.
+    ///
+    /// Returns a [`GridFsStream`] to which the application will write the contents.
+    pub fn open_upload_stream_with_id<T>(
+        &self,
+        id: T,
+        filename: String,
+        options: GridFsUploadOptions,
+    ) -> GridFsStream {
+        todo!()
+    }
+
+    /// Uploads a user file to a GridFS bucket. The application supplies a custom file id.
+    pub fn upload_from_stream_with_id<T>(
+        &self,
+        id: T,
+        filename: String,
+        source: GridFsStream,
+        option: GridFsUploadOptions,
+    ) {
+        todo!()
+    }
+
+    /// Opens and returns a [`GridFsStream`] from which the application can read
+    /// the contents of the stored file specified by `id`.
+    pub fn open_download_stream<T>(&self, id: T) -> GridFsStream {
+        todo!()
+    }
+
+    /// Opens and returns a [`GridFsStream`] from which the application can read
+    /// the contents of the stored file specified by `filename` and the revision
+    /// in `options`.
+    pub fn open_download_stream_by_name(
+        &self,
+        filename: String,
+        options: GridFsDownloadByNameOptions,
+    ) -> GridFsStream {
+        todo!()
+    }
+
+    /// Downloads the contents of the stored file specified by `id` and writes
+    /// the contents to the destination [`GridFsStream`].
+    pub fn download_to_stream<T>(&self, id: T, destination: GridFsStream) {
+        todo!()
+    }
+
+    /// Downloads the contents of the stored file specified by `filename` and by
+    /// the revision in `options` and writes the contents to the destination
+    /// [`GridFsStream`].
+    pub fn download_to_stream_by_name(
+        &self,
+        filename: String,
+        destination: GridFsStream,
+        options: GridFsDownloadByNameOptions,
+    ) {
+        todo!()
+    }
+
+    /// Given an `id`, deletes the stored file's files collection document and
+    /// associated chunks from a [`GridFsBucket`].
+    pub fn delete<T>(&self, id: T) {
+        todo!()
+    }
+
+    /// Finds and returns the files collection documents that match the filter.
+    pub fn find<T>(&self, filter: Document, options: GridFsBucketOptions) -> Result<Cursor<T>> {
```

Comment:
The nice thing about using the `FilesCollectionDocument` struct rather than a plain `Document` is that the validation of the data returned is done ahead of time by the cursor, so a user wanting to inspect a field just needs to access it on the struct. With a `Document`, a user wanting to inspect the `length` field would need to do something like:
```
doc.get_i64("length").unwrap();
```
which requires knowing what the type of that field should be. Even though the unwrap should never fail it's nice to have the guarantee of the type system rather than needing to assume an invariant/write extra error handling.

That said, I'd be interested in what other members of the team think about using a defined struct here; we can leave this thread open for their input once I tag them in for review.

Reply:
I like the suggestion to have a `FilesCollectionDocument` type, for the reasons Isabel mentioned; it's a well-defined format and when possible we generally try to model those types